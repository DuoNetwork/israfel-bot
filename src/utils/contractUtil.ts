import moment from 'moment';
import Web3 from 'web3';
import DualClassWrapper from '../../../duo-contract-wrapper/src/DualClassWrapper';
import Web3Wrapper from '../../../duo-contract-wrapper/src/Web3Wrapper';
import Web3Util from '../../../israfel-relayer/src/utils/Web3Util';
import * as CST from '../common/constants';
import { IAccounts, IDualClassStates, IOption } from '../common/types';
import priceUtil from './priceUtil';
import util from './util';

const Tx = require('ethereumjs-tx');

export class ContractUtil {
	public dualClassCustodianWrapper: DualClassWrapper;
	public web3Wrapper: Web3Wrapper | null = null;
	public web3Util: Web3Util;
	public web3: Web3;

	constructor(web3Util: Web3Util, web3Wrapper: Web3Wrapper, option: IOption) {
		this.web3Wrapper = web3Wrapper;
		this.web3Util = web3Util;
		this.dualClassCustodianWrapper = new DualClassWrapper(
			web3Wrapper,
			web3Wrapper.contractAddresses.Custodians.Beethoven[option.tenor].custodian.address
		);
		this.web3 = new Web3(
			option.source
				? new Web3.providers.HttpProvider(option.provider)
				: new Web3.providers.WebsocketProvider(option.provider)
		);
	}

	public async estimateBeethovenPrice(option: IOption): Promise<number[]> {
		if (!this.dualClassCustodianWrapper) {
			util.logDebug(`no dualClassWrapper initiated`);
			return [];
		}

		const states: IDualClassStates = await this.dualClassCustodianWrapper.getStates();

		const price = await priceUtil.getLastPrice(15, option);

		if (!price) {
			util.logDebug(`no price calculated`);
			return [];
		}

		const resetTime = Math.floor(states.resetPriceTime / 1000);
		const { resetPrice, beta, alpha, periodCoupon } = states;

		const time = Math.floor(moment.utc().valueOf() / 1000);

		const numOfPeriods = Math.floor((time - resetTime) / (states.period / 1000));
		const navParent = (price / resetPrice / beta) * (1 + alpha);

		const navA = 1 + numOfPeriods * Number(periodCoupon);
		const navAAdj = navA * alpha;
		if (navParent <= navAAdj) return [navParent / alpha, 0];
		else return [navA / price, navParent - navAAdj / price];
	}

	public async estimateBeethovenTokenCreateAmt(ethAmount: number): Promise<number[]> {
		if (!this.dualClassCustodianWrapper || ethAmount <= 0) {
			util.logDebug(`no dualClassWrapper initiated`);
			return [];
		}
		const states: IDualClassStates = await this.dualClassCustodianWrapper.getStates();
		const tokenValueB =
			(((1 - states.createCommRate) * states.resetPrice) / (1 + states.alpha)) * ethAmount;
		const tokenValueA = states.alpha * tokenValueB;
		return [tokenValueA, tokenValueB];
	}

	private getMainAccount(): IAccounts {
		const faucetAccount = require('../keys/faucetAccount.json');

		return {
			address: faucetAccount.publicKey,
			privateKey: faucetAccount.privateKey
		};
	}

	private async ethTransferRaw(
		web3: Web3,
		from: string,
		privatekey: string,
		to: string,
		amt: number,
		nonce: number
	) {
		const rawTx = {
			nonce: nonce,
			gasPrice: web3.utils.toHex((await web3.eth.getGasPrice()) || CST.DEFAULT_GAS_PRICE),
			gasLimit: web3.utils.toHex(23000),
			from: from,
			to: to,
			value: web3.utils.toHex(web3.utils.toWei(amt.toPrecision(3) + '', 'ether'))
		};
		return web3.eth
			.sendSignedTransaction('0x' + this.signTx(rawTx, privatekey))
			.then(receipt => util.logInfo(JSON.stringify(receipt, null, 4)));
	}

	public signTx(rawTx: object, privateKey: string): string {
		try {
			const tx = new Tx(rawTx);
			tx.sign(new Buffer(privateKey, 'hex'));
			return tx.serialize().toString('hex');
		} catch (err) {
			util.logError(err);
			return '';
		}
	}

	public async checkBalance(
		pair: string,
		tokenIndex: number,
		addresses: string[]
	): Promise<string[]> {
		const [code1, code2] = pair.split('|');

		for (const address of addresses)
			if (this.web3Util) {
				// ethBalance
				const ethBalance = await this.web3Util.getEthBalance(address);
				util.logInfo(`the ethBalance of ${address} is ${ethBalance}`);
				if (ethBalance < CST.MIN_ETH_BALANCE) {
					util.logDebug(
						`the address ${address} current eth balance is ${ethBalance}, make transfer...`
					);
					addresses = addresses.filter(addr => addr !== address);
					const faucetAccount: IAccounts = this.getMainAccount();

					await this.ethTransferRaw(
						this.web3,
						faucetAccount.address,
						faucetAccount.privateKey,
						address,
						CST.MIN_ETH_BALANCE,
						await this.web3Util.getTransactionCount(faucetAccount.address)
					);
					continue;
				}

				// wEthBalance
				const wEthBalance = await this.web3Util.getTokenBalance(code2, address);
				if (wEthBalance < CST.MIN_WETH_BALANCE) {
					util.logInfo(
						`the address ${address} current weth balance is ${wEthBalance}, wrapping...`
					);
					const amtToWrap = CST.MIN_WETH_BALANCE + 0.1 - wEthBalance;
					if (ethBalance > amtToWrap) {
						util.logInfo(`start wrapping for ${address} with amt ${amtToWrap}`);
						await this.web3Util.wrapEther(amtToWrap, address);
					} else {
						util.logInfo(
							`the address ${address} current weth balance is ${wEthBalance}, not enought eth to wrapping, skip...`
						);
						addresses = addresses.filter(addr => addr !== address);
						continue;
					}
				}

				// tokenBalance
				const tokenBalance = await this.web3Util.getTokenBalance(code1, address);
				util.logInfo(`the ${code1} tokenBalance of ${address} is ${tokenBalance}`);
				const accountsBot: IAccounts[] = require('../keys/accountsBot.json');
				const account = accountsBot.find(a => a.address === address);
				const gasPrice = Math.max(
					await this.web3Util.getGasPrice(),
					CST.DEFAULT_GAS_PRICE * Math.pow(10, 9)
				);
				if (tokenBalance < CST.MIN_TOKEN_BALANCE) {
					util.logInfo(
						`the address ${address} current token balance of ${code1} is ${tokenBalance}, need create more tokens...`
					);

					const tokenValues = await this.estimateBeethovenTokenCreateAmt(
						ethBalance - CST.MIN_ETH_BALANCE - 0.1
					);
					if (
						tokenValues.length &&
						tokenValues[tokenIndex] + tokenBalance > CST.MIN_TOKEN_BALANCE
					) {
						util.logDebug(`creating token ${code1}`);
						if (account)
							await this.dualClassCustodianWrapper.createRaw(
								address,
								account.privateKey,
								gasPrice,
								CST.CREATE_GAS,
								Number(
									Number(
										(ethBalance - CST.MIN_ETH_BALANCE - 0.1).toFixed(10)
									).toPrecision(3)
								)
							);
						else {
							util.logDebug(`the address ${address} cannot create, skip...`);
							addresses = addresses.filter(addr => addr !== address);
							continue;
						}
					}
				} else if (tokenBalance >= CST.MAX_TOKEN_BALANCE) {
					util.logDebug(
						`the address ${address} current token balance of ${code1} is ${tokenBalance}, need redeem back...`
					);
					if (account) {
						const states: IDualClassStates = await this.dualClassCustodianWrapper.getStates();
						await this.dualClassCustodianWrapper.redeemRaw(
							address,
							account.privateKey,
							tokenBalance - CST.MAX_TOKEN_BALANCE,
							(tokenBalance - CST.MAX_TOKEN_BALANCE) / states.alpha,
							gasPrice,
							CST.REDEEM_GAS
						);
					}
				}

				// tokenAllowance
				const tokenAllowance = await this.web3Util.getProxyTokenAllowance(code1, address);
				if (tokenAllowance <= 0) {
					util.logDebug(`the address ${address} token allowance is 0, approvaing.....`);
					this.web3Util.setUnlimitedTokenAllowance(code1, address);
				}
			}

		return addresses;
	}
}
