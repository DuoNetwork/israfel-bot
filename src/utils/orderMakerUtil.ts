// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import WebSocket from 'ws';
import orderUtil from '../../../israfel-relayer/src/utils/orderUtil';
import Web3Util from '../../../israfel-relayer/src/utils/Web3Util';
import * as CST from '../common/constants';
import { IOrderBookSnapshotLevel, IWsAddOrderRequest } from '../common/types';
import { ContractUtil } from './contractUtil';
import util from './util';

export class OrderMakerUtil {
	public web3Util: Web3Util | null = null;
	public ws: WebSocket | null = null;
	public availableAddrs: string[] = [];
	public currentAddrIdx: number = 0;
	public contractUtil: ContractUtil | null = null;
	constructor(web3Util: Web3Util, ws: WebSocket, contractUtil: ContractUtil) {
		this.web3Util = web3Util;
		this.ws = ws;
		this.contractUtil = contractUtil;
	}

	public getCurrentAddress() {
		const currentAddr = this.availableAddrs[this.currentAddrIdx];
		this.currentAddrIdx = (this.currentAddrIdx + 1) % this.availableAddrs.length;
		return currentAddr;
	}

	public async placeOrder(isBid: boolean, price: number, amount: number, pair: string) {
		if (!this.web3Util) throw new Error('no web3Util initiated');
		if (!this.web3Util.isValidPair(pair)) throw new Error('invalid pair');
		const [code1, code2] = pair.split('|');
		const token1 = this.web3Util.getTokenByCode(code1);
		if (!token1) throw new Error('invalid pair');
		const address1 = token1.address;
		const address2 = this.web3Util.getTokenAddressFromCode(code2);

		const amountAfterFee = orderUtil.getAmountAfterFee(
			amount,
			price,
			token1.feeSchedules[code2],
			isBid
		);

		const expiry = Math.floor(util.getExpiryTimestamp(false) / 1000);

		if (!amountAfterFee.makerAssetAmount || !amountAfterFee.takerAssetAmount)
			throw new Error('invalid amount');

		console.log('start creating rawOrder ############');

		const rawOrder = await this.web3Util.createRawOrder(
			pair,
			this.getCurrentAddress(),
			isBid ? address2 : address1,
			isBid ? address1 : address2,
			amountAfterFee.makerAssetAmount,
			amountAfterFee.takerAssetAmount,
			expiry
		);

		console.log(rawOrder);

		const msg: IWsAddOrderRequest = {
			method: CST.DB_ADD,
			channel: CST.DB_ORDERS,
			pair: pair,
			orderHash: rawOrder.orderHash,
			order: rawOrder.signedOrder
		};
		if (!this.ws) {
			console.log('no client initiated');
			return;
		}

		util.logInfo(
			'send add order request' +
				JSON.stringify({
					price: price,
					amount: amount,
					isBid: isBid
				})
		);
		this.ws.send(JSON.stringify(msg));
	}

	private async createBeethovenOrderBook(
		pair: string,
		isBid: boolean,
		contractType: string,
		contractTenor: string,
		midPrice: number,
		totalSize: number,
		numOfOrders: number
	) {
		if (!this.contractUtil) {
			util.logDebug(`no contractUtil initiated`);
			return;
		}
		util.logInfo(`make depth for ${contractType} contract`);
		console.log(contractTenor);
		if (![CST.TENOR_PPT, CST.TENOR_M19].includes(contractTenor)) {
			util.logInfo('wrong contract tenor');
			return;
		}
		util.logInfo(`make depth for ${contractTenor} contract`);
		const amountPerLevel = totalSize / numOfOrders;

		for (let i = 0; i < numOfOrders; i++) {
			const bidPrice = util.round(midPrice - (i + 1) * CST.PRICE_STEP, '3');
			const askPrice = util.round(midPrice + (i + 1) * CST.PRICE_STEP, '3');
			const bidAmt = util.round(amountPerLevel + Math.random() * 10, '1');
			const askAmt = util.round(amountPerLevel + Math.random() * 10, '1');

			util.logInfo(
				`placing an ${isBid ? 'bid' : 'ask'} order, with price ${
					isBid ? bidPrice : askPrice
				} with amount ${isBid ? bidAmt : askAmt}`
			);
			await this.placeOrder(
				isBid,
				isBid ? bidPrice : askPrice,
				isBid ? bidAmt : askAmt,
				pair
			);
			util.sleep(1000);
			break;
		}
	}

	public async takeOneSideOrders(
		pair: string,
		isBid: boolean,
		orderBookSide: IOrderBookSnapshotLevel[]
	) {
		for (const orderLevel of orderBookSide) {
			util.logDebug(
				`taking an  ${isBid ? 'bid' : 'ask'} order with price ${orderLevel.price} amount ${
					orderLevel.balance
				}`
			);
			await this.placeOrder(!isBid, orderLevel.price, orderLevel.balance, pair);
			util.sleep(1000);
		}
	}

	public async createOrderBookSide(
		pair: string,
		isBid: boolean,
		contractType: string,
		contractTenor: string,
		midPrice: number,
		totalSize: number,
		numOfOrders: number
	) {
		if (!this.contractUtil) {
			util.logDebug(`no contractUtil initiated`);
			return;
		}
		util.logInfo(`start making side for ${contractType} ${isBid ? 'bid' : 'ask'}`);

		switch (contractType) {
			case CST.BEETHOVEN:
				await this.createBeethovenOrderBook(
					pair,
					isBid,
					contractType,
					contractTenor,
					midPrice,
					totalSize,
					numOfOrders
				);
				break;
			default:
				util.logDebug(`incorrect contract type specified`);
				return;
		}
	}
}
