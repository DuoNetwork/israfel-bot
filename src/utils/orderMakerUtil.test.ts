// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import Web3Wrapper from '../../../duo-contract-wrapper/src/Web3Wrapper';
import Web3Util from '../../../israfel-relayer/src/utils/Web3Util';
import { ContractUtil } from './contractUtil';
import { OrderMakerUtil } from './orderMakerUtil';
import util from './util';
const mnemonic = require('../keys/mnemomic.json');

const web3Util = new Web3Util(null, false, mnemonic.mnemomic, false);
const option = {
	live: false,
	source: 'infura',
	provider: 'provider',
	token: 'token',
	type: 'Beethoven',
	tenor: 'Perpetual'
} as any;
const web3Wrapper = new Web3Wrapper(null, option.source, option.provider, option.live);
const contractUtil = new ContractUtil(web3Util, web3Wrapper, option);

test('createDualTokenOrderBook wrong tenor', async () => {
	const orderMakerUtil = new OrderMakerUtil(web3Util, contractUtil);
	orderMakerUtil.placeOrder = jest.fn(() => Promise.resolve());
	util.sleep = jest.fn(() => Promise.resolve());
	await orderMakerUtil.createDualTokenOrderBook({
		pair: 'pair',
		isBid: true,
		contractTenor: 'tenor',
		midPrice: 120,
		totalSize: 50,
		numOfOrders: 3,
		existingPriceLevel: []
	});
	expect(util.sleep as jest.Mock).not.toBeCalled();
	expect(orderMakerUtil.placeOrder as jest.Mock).not.toBeCalled();
});

test('createDualTokenOrderBook', async () => {
	const orderMakerUtil = new OrderMakerUtil(web3Util, contractUtil);
	orderMakerUtil.placeOrder = jest.fn(() => Promise.resolve(true));
	util.sleep = jest.fn(() => Promise.resolve());
	await orderMakerUtil.createDualTokenOrderBook({
		pair: 'pair',
		isBid: true,
		contractTenor: 'M19',
		midPrice: 0.0012,
		totalSize: 50,
		numOfOrders: 3,
		existingPriceLevel: []
	});
	expect((orderMakerUtil.placeOrder as jest.Mock).mock.calls).toMatchSnapshot();
});

test('takeOneSideOrder', async () => {
	const orderMakerUtil = new OrderMakerUtil(web3Util, contractUtil);
	orderMakerUtil.placeOrder = jest.fn(() => Promise.resolve(true));
	util.sleep = jest.fn(() => Promise.resolve());
	await orderMakerUtil.takeOneSideOrders('pair', true, [
		{
			price: 0.0012,
			balance: 20,
			count: 10
		},
		{
			price: 0.0011,
			balance: 10,
			count: 5
		}
	]);
	expect((orderMakerUtil.placeOrder as jest.Mock).mock.calls).toMatchSnapshot();
});

// test('placeOrder', async () => {
// 	const orderMakerUtil = new OrderMakerUtil(null as any, contractUtil);
// 	orderMakerUtil.web3Util = null as any;
// 	expect(() => orderMakerUtil.placeOrder(true, 0.0012, 10, 'pair')).toThrowError();
// });
