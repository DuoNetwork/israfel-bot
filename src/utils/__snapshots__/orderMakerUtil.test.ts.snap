// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`takeOneSideOrder 1`] = `
Array [
  Array [
    false,
    0.0012,
    20,
    "pair",
  ],
  Array [
    false,
    0.0011,
    10,
    "pair",
  ],
]
`;
